### PR TITLE
Add .NET Core Framework

### PR DESCRIPTION
**Pull request by rprouse**
**Ready to merge?**

This adds support for .NET Core/ASP.NET 5 and Windows 10 RT. I am creating a PR now so that we can discuss.

Remaining work,

- Use the nunitlite package to create a project to run the tests and fix any errors
- Add the new assemblies to the NuGet packages
- Update nunit.proj to build

I am currently building this on Windows 8.1, but this will make Visual Studio 2015 required to build.

### CODE REVIEW

**Comment by rprouse** [comment]
We don't want to spend too much time on getting .NET Core support in for 3.0, but I wanted to take advantage of the help that the Microsoft guys provided last week. This code builds and all tests are passing. I propose that I remove the .NET Core projects from the solution and we check in these changes which will allow us to add proper .NET Core support after 3.0 is released. A fair amount of work went into this, so I would hate to lose that.

I will make a few comments in the code diffs to explain why the changes were needed.

File: src/Common/nunit/Options.cs [hunk 1]
```diff
@@ -533,6 +546,17 @@ public OptionSet (Converter<string, string> localizer)
         public Converter<string, string> MessageLocalizer {
             get {return localizer;}
         }
+#else
+        string localizer(string msg)
```
Comment (rprouse) [inline src/Common/nunit/Options.cs]: There is no Convert class in DNX and we don't use the localizer, so creating straight pass-through methods for this and the MessageLocalizer is the equivalent to the code in the default constructor that we use elsewhere.

File: src/NUnitFramework/framework/Attributes/RandomAttribute.cs [hunk 2]
```diff
@@ -176,7 +178,7 @@ public IEnumerable GetData(IParameterInfo parameter)
                     _source = new SByteDataSource(_count);
                 else if (parmType == typeof(decimal))
                     _source = new DecimalDataSource(_count);
-                else if (parmType.IsEnum)
+                else if (parmType.GetTypeInfo().IsEnum)
```
Comment (rprouse) [inline src/NUnitFramework/framework/Attributes/RandomAttribute.cs]: The majority of the changes are of this nature. Many of the properties and methods that we used to find directly on type are now accessed through the TypeInfo. On older platforms, I created an extension method for this on Type that simply returns the type.

File: src/NUnitFramework/framework/Compatibility/ReflectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        /// <summary>
+        /// Returns an array of custom attributes of the specified type applied to this member
+        /// </summary>
+        /// <remarks> Portable throws an argument exception if T does not
+        /// derive from Attribute. NUnit uses interfaces to find attributes, thus
+        /// this method</remarks>
+        public static IEnumerable<T> GetAttributes<T>(this MemberInfo info, bool inherit) where T : class
+        {
+            return GetAttributesImpl<T>(info.GetCustomAttributes(inherit));
```
Comment (rprouse) [inline src/NUnitFramework/framework/Compatibility/ReflectionExtensions.cs]: In DNX, GetCustomAttributes throws an argument exception if the type is not derived from Attribute. NUnit searches for attributes based on the interface they implement, so these extension methods get all attributes and find the ones that are assignable from T.

File: src/NUnitFramework/framework/Constraints/PropertyConstraint.cs [hunk 4]
```diff
@@ -62,11 +62,7 @@ public override ConstraintResult ApplyTo<TActual>(TActual actual)
                 actualType = actual.GetType();
 
             PropertyInfo property = actualType.GetProperty(name,
-#if PORTABLE
                 BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
-#else
-                BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.GetProperty);
```
Comment (rprouse) [inline src/NUnitFramework/framework/Constraints/PropertyConstraint.cs]: BindingFlags.GetProperty is not available in Portable or DNX, but it was also not being used correctly here. GetProperty is used to get the value of the property, not to get the getter property.

File: src/NUnitFramework/framework/Internal/MethodWrapper.cs [hunk 5]
```diff
@@ -165,15 +170,23 @@ public IMethodInfo MakeGenericMethod(params Type[] typeArguments)
...
-        /// <summary>
-        /// Gets a value indicating whether one or more attributes of the spcified type are defined on the method.
-        /// </summary>
+            /// <summary>
+            /// Gets a value indicating whether one or more attributes of the spcified type are defined on the method.
+            /// </summary>
         public bool IsDefined<T>(bool inherit)
         {
+#if NETCORE
+            return MethodInfo.GetCustomAttributes(inherit).Any(a => typeof(T).IsAssignableFrom(a.GetType()));
```
Comment (rprouse) [inline src/NUnitFramework/framework/Internal/MethodWrapper.cs]: IsDefined also throws if T does not derive from Attribute, thus the custom implementation.

File: src/NUnitFramework/framework/nunit.framework-netcore/nunit.framework-netcore.csproj [hunk 6]
```diff
@@ -0,0 +1,1062 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (rprouse) [inline src/NUnitFramework/framework/nunit.framework-netcore/nunit.framework-netcore.csproj]: This is another odd thing about DNX. If the project is in the same directory as our other projects, Visual Studio uses the project.json file that gets added to DNX projects for all the projects in that directory causing all the other projects to fail to build. As a workaround, I put the projects in subfolders.

File: src/NUnitFramework/nunitlite.runner/nunitlite.runner-netcore/Runner/AutoRun.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
+﻿// ***********************************************************************
```
Comment (rprouse) [inline src/NUnitFramework/nunitlite.runner/nunitlite.runner-netcore/Runner/AutoRun.cs]: A couple of these runner classes were getting really complicated and hard to follow with the defines for all the different platforms, so it was easier with a couple to make copies, clean them up and rewrite them just for DNX.

File: src/NUnitFramework/tests/Attributes/RangeAttributeTests.cs [hunk 8]
```diff
@@ -408,7 +412,11 @@ private void CheckValues(string methodName, params object[] expected)
         {
             var method = GetType().GetMethod(methodName, BindingFlags.NonPublic | BindingFlags.Instance);
             var param = method.GetParameters()[0];
+#if NETCORE
+            var attr = param.GetCustomAttributes(typeof(ValuesAttribute), false).First() as ValuesAttribute;
```
Comment (rprouse) [inline src/NUnitFramework/tests/Attributes/RangeAttributeTests.cs]: GetCustomAttributes returns IEnumerable on DNX. This was one of the few places that caused a problem.

File: src/NUnitFramework/tests/Internal/UnhandledExceptionTests.cs [hunk 9]
```diff
@@ -34,7 +34,7 @@ namespace NUnit.Framework.Tests
     [TestFixture, Explicit("These tests fail by design")]
     public class UnhandledExceptionTests
     {
-        #region Normal
+#region Normal
```
Comment (rprouse) [inline src/NUnitFramework/tests/Internal/UnhandledExceptionTests.cs]: This needs fixing.

File: src/NUnitFramework/framework/Internal/MethodWrapper.cs [hunk 10]
```diff
@@ -165,15 +170,23 @@ public IMethodInfo MakeGenericMethod(params Type[] typeArguments)
...
+            return MethodInfo.GetAttributes<T>(inherit).ToArray();
+#else
             return (T[])MethodInfo.GetCustomAttributes(typeof(T), inherit);
+#endif
         }
 
-        /// <summary>
-        /// Gets a value indicating whether one or more attributes of the spcified type are defined on the method.
-        /// </summary>
+            /// <summary>
```
Comment (rprouse) [inline src/NUnitFramework/framework/Internal/MethodWrapper.cs]: I will fix the extra indenting here.

**Comment by rprouse** [comment]
I have removed the DNX projects from the main NUnit solution and created a separate DNX solution.

If everyone is okay with these changes and the direction, then this is ready to merge. It is a large diff, but most of it is just rote changes.

**Comment by oznetmaster** [comment]
I am curious.  Currently, all of the source files seem to have their #region/#endregion statements indented.

This PR changes them all to be unindented.

Is this an intention coding standard change?

**Comment by rprouse** [comment]
No, whenever I entered an #endif in Visual Studio, it unindented all the regions. I fixed wherever I noticed, but I guess I missed some. Just point them out and I will revert them. I did a scan of the diffs and thought I got them all.

File: src/Common/nunit/CommandLineOptions.cs [hunk 11]
```diff
@@ -62,9 +66,9 @@ public CommandLineOptions(params string[] args)
                 Parse(args);
         }
         
-        #endregion
+#endregion
```
Comment (rprouse) [inline src/Common/nunit/CommandLineOptions.cs]: Regions in this file need to be fixed.

**Comment by CharliePoole** [comment]
I'd like to review this before we merge it. Since we are working directly in master, we need ot make a careful decision about whether we want this in the 3.0 release or not. Up to now, we have been treating #575 as a post-3.0 issue.

**Comment by rprouse** [comment]
@CharliePoole, the review should be fairly straight-forward. I know that #575 is a post-3.0 issue and I think we should leave it that way, but this makes the code changes necessary to allow us to support several platforms more easily after 3.0.

**Comment by CharliePoole** [comment]
@rprouse Fair enough. I'll get to it. :-)

**Comment by CharliePoole** [comment]
I'm surprised to see that the .NET core build has a command-line at all. Is the runner .NET core as well as the framework?

FWIW, I'm a little scared to merge all this. :-)

**Comment by rprouse** [comment]
Yes,. Net core has a command line.

What parts of this are making you nervous? Once you exclude the typeinfo
extension, there isn't much else that changes the main frameworks. Most of
the rest are #if NETCORE.

It might help to add comments to the areas that concern you so I can
explain.

If we want to support Windows 10 or. NET Core, we are going to have to make
changes like this sooner or later. It was also useful to see that the
branch from the Microsoft guy was nearly the same as what I did. I ended up
combining the best ideas from both of our branches.

**Comment by CharliePoole** [comment]
OK, that explains why we still have command-line options. :-) Oddly, we
also have them for Silverlight, where we have no way to use them.

The changes all make sense. I'm only worried about conflicts, given the
size. I plan to try to merge it on top of the other Beta-5 stuff. Should be
doable.

**Comment by rprouse** [comment]
Looks like the beta is delayed until the end of the week. I will try to get latest from master merged into this branch so there are no conflicts before then.

**Comment by richlander** [comment]
@rprouse Sweet! Anything more we can do to help you? We'd love to see this available sooner than later.

**Comment by CharliePoole** [comment]
Key question... do you want to release a build for .NET core or just have
the source ready for it?

**Comment by rprouse** [comment]
At the least, I want to get the code ready. If we could get it in the nuget
for the RC, then that would be even better. That would give us a really
good foundation to build upon. DNX is very NuGet based, so having the
packages available will make out of band development of templates and
runners much easier.

I also want to take advantage of the help from Microsoft while it is
available. It is also one of our most discussed and linked to issues, so
feel it deserves some attention.

All tests are running and passing for supported platforms. The DNX
framework needs testing and possibly more work, but getting this pr merged
will get the large changes behind us and tested earlier making future
changes much smaller and manageable.

File: nunit.sln [hunk 12]
```diff
@@ -1,6 +1,8 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2012
+# Visual Studio 14
```
Comment (CharliePoole) [inline nunit.sln]: It always seems strange to me that simply opening a solution in a newer version of VS causes changes to be made!

File: src/Common/nunit/CommandLineOptions.cs [hunk 13]
```diff
@@ -40,7 +40,9 @@ namespace NUnit.Common
     public class CommandLineOptions : OptionSet
     {
         private bool validated;
+#if !NETCORE
```
Comment (CharliePoole) [inline src/Common/nunit/CommandLineOptions.cs]: It would help to have a summary somewhere of what options are unavailable in NETCORE and why.

File: nunit.sln (hunk 12 above)
Comment (rprouse) [inline nunit.sln]: Especially since those changes are just comments. I think it only happens
if you change the solution. I had included the DNX projects, then backed
out those changes in a later commit.

File: src/Common/nunit/Options.cs [hunk 14]
```diff
@@ -349,17 +353,25 @@ public string[] GetValueSeparators ()
         protected static T Parse<T> (string value, OptionContext c)
         {
             Type tt = typeof (T);
+#if NETCORE
+            bool nullable = tt.GetTypeInfo().IsValueType && tt.GetTypeInfo().IsGenericType && 
+                !tt.GetTypeInfo().IsGenericTypeDefinition && 
+                tt.GetGenericTypeDefinition () == typeof (Nullable<>);
+            Type targetType = nullable ? tt.GetGenericArguments () [0] : typeof (T);
+#else
```
Comment (CharliePoole) [inline src/Common/nunit/Options.cs]: I see you had to do this to compile. Since we don't do actually use any Nullable  types in our options, it makes me think we should eventually just build our own version of Mono.Options that has what we need and no more.

File: src/Common/nunit/Options.cs (hunk 1 above)
Comment (CharliePoole) [inline src/Common/nunit/Options.cs]: Another point for my comment above about taking this class over in the future.

File: src/NUnitFramework/framework/Attributes/RandomAttribute.cs (hunk 2 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Attributes/RandomAttribute.cs]: I used the name TypeInfo in several places in my Type wrapper. Maybe we should change that (some time) to avoid confusion with the "real" TypeInfo.

File: src/NUnitFramework/framework/Compatibility/AttributeHelper.cs [hunk 15]
```diff
@@ -64,12 +64,16 @@ public static Attribute[] GetCustomAttributes(object actual, Type attributeType,
             var assembly = actual as Assembly;
             if (assembly != null)
             {
+#if NETCORE
+                return (Attribute[])assembly.GetCustomAttributes(attributeType);
+#else
                 return (Attribute[])assembly.GetCustomAttributes(attributeType, inherit);
+#endif
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Compatibility/AttributeHelper.cs]: Why not just add an extension method?

File: src/NUnitFramework/nunitlite.runner/nunitlite.runner-netcore/Runner/AutoRun.cs (hunk 7 above)
Comment (CharliePoole) [inline src/NUnitFramework/nunitlite.runner/nunitlite.runner-netcore/Runner/AutoRun.cs]: Let's try this and see if it may be a better way to go for other runners as well.

File: src/NUnitFramework/framework/Internal/ParameterWrapper.cs [hunk 16]
```diff
@@ -24,8 +24,13 @@
 using System;
 using System.Collections.Generic;
 using System.Reflection;
+using NUnit.Framework.Compatibility;
 using NUnit.Framework.Interfaces;
 
+#if NETCORE
+using System.Linq;
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/ParameterWrapper.cs]: I see it saves a line or two of code to use Linq, but it adds another #if, which seems like a poor tradeoff. I noticed this in a few other places as well.

**Comment by CharliePoole** [comment]
This looks OK to merge, with or without some of my suggestions right now.

I see that you used a separate solution, which I like because it avoids forcing developers to use a newer version of Visual Studio if they are not working on this new build.

I'd additionally like to see some sort of summary of options not supported by the runner, which you could put here as a comment and which we would eventually paste into the docs.

File: nunit.sln (hunk 12 above)
Comment (CharliePoole) [inline nunit.sln]: Right - although others are not comments. They seem harmless however.

File: src/Common/nunit/CommandLineOptions.cs (hunk 13 above)
Comment (rprouse) [inline src/Common/nunit/CommandLineOptions.cs]: I agree, same for NUnitLite. My plan was to start small and once we have time, create a more full featured runner. I'm still not sure if it will have to be a full console runner, an nunitlite type runner or something else. So far, I am adding an nunitlite type runner.

File: src/Common/nunit/Options.cs (hunk 14 above)
Comment (rprouse) [inline src/Common/nunit/Options.cs]: Another agreement :smile: 

There is a lot we don't use and this has diverged so much from the original, it would be nice to clean it up.

**Comment by rprouse** [comment]
@CharliePoole, I've made a couple of small changes based on your code review. I will include the command line options that we support once that firms up with following pull requests.

**Comment by CharliePoole** [comment]
@Coveralls It's not clear what your comment on this closed issue means. Are you identifying a new problem?

**Comment by mikkelbu** [comment]
@CharliePoole coveralls is an online tool for code coverage, so you will probably not get an answer. Why it has run now, and on this particular commit, I don't know.

**Comment by CharliePoole** [comment]
@mikkelbu Maybe they are testing something. Rather invasive if you ask me.

**Comment by CharliePoole** [comment]
@mikkelbu LOL! I thought somebody had chosen a garment worn by farmers and mechanics as their alias!
